File: frontend/src/constants/estados.js
// Vocabulario canonico de estados de una orden. Espejo del backend
// (backend/app/core/estados.py) — misma key, label y color, para que el
// donut/badges del dashboard (que ya vienen calculados del backend) y
// las vistas que crean/editan ordenes (que todavia mandan el label como
// texto libre) queden siempre consistentes.
//
// Antes de este archivo esta lista estaba duplicada (con variaciones
// sutiles) en StatusBadge.vue, OrderCard.vue, DashboardView.vue y los
// selects de OrdenesView.vue/OrdenDetalleView.vue. Ahora todos importan
// de aca.

export const ESTADOS = [
  { key: 'pendiente', label: 'Pendiente', color: '#64748b', icon: 'clock' },
  { key: 'diagnostico', label: 'Diagnostico', color: '#0ea5e9', icon: 'stethoscope' },
  { key: 'esperando_repuesto', label: 'Esperando repuesto', color: '#f97316', icon: 'repuesto' },
  { key: 'reparacion', label: 'En reparacion', color: '#3b66f5', icon: 'wrench' },
  { key: 'listo', label: 'Listo', color: '#22c55e', icon: 'check' },
  { key: 'entregado', label: 'Entregado', color: '#a855f7', icon: 'check' },
  { key: 'cancelado', label: 'Cancelado', color: '#ef4444', icon: 'trash' },
]

// Labels en el orden del flujo de trabajo, listos para usar en un
// <select>. Mismo string que guarda el backend en Orden.estado.
export const ESTADOS_LABELS = ESTADOS.map((estado) => estado.label)

const ALIASES = {
  pendiente: ['pendiente', 'recibido'],
  diagnostico: ['diagnostico'],
  esperando_repuesto: ['repuesto'],
  reparacion: ['reparacion', 'proceso'],
  listo: ['listo', 'reparad'],
  entregado: ['entreg'],
  cancelado: ['cancel'],
}

/**
 * Resuelve el string libre de una orden (ej. "En reparacion", o
 * variantes legacy como "recibido") a su definicion canonica. Mismo
 * criterio de matching por substring que usa el backend, para que un
 * dato historico se vea igual en toda la app.
 */
export function resolveEstado(raw) {
  const value = (raw || '').toLowerCase().trim()

  for (const estado of ESTADOS) {
    if (ALIASES[estado.key].some((needle) => value.includes(needle))) {
      return estado
    }
  }

  return ESTADOS[0] // fallback: pendiente
}

File: backend/app/core/estados.py
"""
Vocabulario canonico de estados de una Orden.

Orden.estado sigue siendo un string libre en la base (ver
app/models/orden.py) — no se migra a un enum de Postgres para no romper
filas historicas ni forzar una migracion. Este modulo es la UNICA fuente
de verdad sobre que estados existen, en que orden van, que color les
corresponde y como se agrupan strings legacy/variantes hacia una key
canonica. Antes de este modulo esa logica de matching estaba duplicada
en el frontend (StatusBadge.vue, DashboardView.vue, OrderCard.vue); acá
vive una sola vez del lado del backend, que es quien calcula los
agregados del dashboard.

Se agrega "Diagnostico" como estado nuevo entre "Pendiente" y
"Esperando repuesto" / "En reparacion".
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class EstadoDef:
    key: str        # identificador estable para JSON (frontend, alertas, etc.)
    label: str       # valor humano que se guarda en Orden.estado
    color: str        # hex, usado en donut/badges para que todo coincida


ESTADOS: tuple[EstadoDef, ...] = (
    EstadoDef("pendiente", "Pendiente", "#64748b"),
    EstadoDef("diagnostico", "Diagnostico", "#0ea5e9"),
    EstadoDef("esperando_repuesto", "Esperando repuesto", "#f97316"),
    EstadoDef("reparacion", "En reparacion", "#3b66f5"),
    EstadoDef("listo", "Listo", "#22c55e"),
    EstadoDef("entregado", "Entregado", "#a855f7"),
    EstadoDef("cancelado", "Cancelado", "#ef4444"),
)

ESTADOS_POR_KEY = {estado.key: estado for estado in ESTADOS}

# Substrings (en minuscula) que identifican cada estado a partir del
# string libre guardado en la base, incluyendo variantes legacy:
# "recibido" era el default viejo de la columna antes de este rediseño,
# y "pagado" viene de un bug ya corregido en pago_routes.py que confundia
# "saldo pagado" con el estado operativo de la orden.
_ALIASES: dict[str, tuple[str, ...]] = {
    "pendiente": ("pendiente", "recibido"),
    "diagnostico": ("diagnostico",),
    "esperando_repuesto": ("repuesto",),
    "reparacion": ("reparacion", "proceso"),
    "listo": ("listo", "reparad"),
    "entregado": ("entreg",),
    "cancelado": ("cancel",),
}

# Estados "abiertos": una orden en cualquiera de estos todavia esta en
# proceso en el taller (no se fue con el cliente ni se descarto).
ESTADOS_ABIERTOS: tuple[str, ...] = ("pendiente", "diagnostico", "esperando_repuesto", "reparacion", "listo")


def resolve_estado_key(raw: str | None) -> str:
    """Convierte el string libre de Orden.estado a una key canonica.

    Recorre ESTADOS en orden (que es el orden real del flujo de trabajo)
    y usa matching por substring, igual criterio que ya usaba el
    frontend. Un valor desconocido (dato legacy no contemplado) cae en
    "pendiente" como fallback honesto, mismo comportamiento que tenia
    StatusBadge.vue.
    """
    value = (raw or "").strip().lower()

    for estado in ESTADOS:
        if any(needle in value for needle in _ALIASES[estado.key]):
            return estado.key

    return "pendiente"
